### PR TITLE
Tdd

### PR DESCRIPTION
**Pull request by zhangshyue**
Week 5 test driven development. There are some changes from the previous pull request. I thought the pull request was updated when I merged it but it was actually not so there is one commit that's not been merged into master.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,95 @@
...
+                    break;
+                }
+            }
+            for (String s : allAttendees) {
+                if (item.getAttendees().contains(s)) {
+                    timeRangesIncludeOptional.add(item.getWhen());
+                    break;
+                }
+            }
+        }
```
Comment (wujingya) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I think we can increase efficiency by using a hashSet for allAttendees and requiredAttendees, then checking whether an attendee exists in these two sets will be faster than if they are two lists.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,95 @@
...
+
+        Collection<TimeRange> ignoreOptional = availableTimeRange(timeRangesIngoreOptional, request);
+        Collection<TimeRange> includeOptional = availableTimeRange(timeRangesIncludeOptional, request);
+        if (request.getAttendees().size() == 0) {
+            return includeOptional;
+        }
+        if (includeOptional.size() != 0) {
+            return includeOptional;
+        }
+        return ignoreOptional;
```
Comment (wujingya) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Let's reorder some of these to minimize the number of operations we perform, like:
```
Collection<TimeRange> includeOptional = availableTimeRange(timeRangesIncludeOptional, request);
if (request.getAttendees().size() == 0 || includeOptional.size() != 0) {
  return includeOptional;
}
return availableTimeRange(timeRangesIngoreOptional, request);
```
This way we don't need to calculate ignoreOptional if we don't need it.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,95 @@
...
+    public Collection<TimeRange> availableTimeRange(List<TimeRange> timeRanges, MeetingRequest request) {
+        Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+        int len = timeRanges.size();
+        if (len == 0) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+        if (len == 1) {
+            return Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, timeRanges.get(0).start(), false),
+                TimeRange.fromStartEnd(timeRanges.get(0).end(), TimeRange.END_OF_DAY, true));
+        }
```
Comment (wujingya) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I feel like you shouldn't need this, the len == 1 case should be handled automatically already

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,95 @@
...
+                index += 1;
+            }
+        }
+            List<TimeRange> queries = new ArrayList<>();
+            int start = TimeRange.START_OF_DAY;
+        for (int i = 0; i < timeRanges.size(); i++) {
+            if (timeRanges.get(i).start() - start >= request.getDuration()) {
+                queries.add(TimeRange.fromStartEnd(start, timeRanges.get(i).start(), false));
+            }
+            start = timeRanges.get(i).end();
```
Comment (wujingya) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I feel like you don't need to merge the overlapping time ranges if you change this line to `start = Math.max(start, timeRanges.get(i).end());`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,95 @@
...
+                timeRanges.remove(index - 1);
+                timeRanges.add(index - 1, TimeRange.fromStartEnd(start, end, false));
+                pre = timeRanges.get(index - 1);
+                len -= 1;
+            } else {
+                index += 1;
+            }
+        }
+            List<TimeRange> queries = new ArrayList<>();
+            int start = TimeRange.START_OF_DAY;
```
Comment (wujingya) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Indentation is off

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,71 @@
...
+        }
+
+        List<TimeRange> timeRangesIngoreOptional = new ArrayList<>();
+        List<TimeRange> timeRangesIncludeOptional = new ArrayList<>();
+        HashSet<String> requiredAttendees = new HashSet<String>(request.getAttendees());
+        HashSet<String> allAttendees = new HashSet<String>(request.getAttendees());
+        allAttendees.addAll(request.getOptionalAttendees());
+        for (Event item : events) {
+            for (String s : item.getAttendees()) {
+                    if (requiredAttendees.contains(s)) {
```
Comment (wujingya) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This line should be unindented one level

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,71 @@
...
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+
+        List<TimeRange> timeRangesIngoreOptional = new ArrayList<>();
+        List<TimeRange> timeRangesIncludeOptional = new ArrayList<>();
+        HashSet<String> requiredAttendees = new HashSet<String>(request.getAttendees());
+        HashSet<String> allAttendees = new HashSet<String>(request.getAttendees());
+        allAttendees.addAll(request.getOptionalAttendees());
+        for (Event item : events) {
+            for (String s : item.getAttendees()) {
```
Comment (wujingya) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: How about change `s` to `attendee`? Same for below

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,71 @@
...
+                if (allAttendees.contains(s)) {
+                    timeRangesIncludeOptional.add(item.getWhen());
+                    break;
+                }
+            }
+        }
+
+        Collection<TimeRange> includeOptional = availableTimeRange(timeRangesIncludeOptional, request);
+        if (request.getAttendees().size() == 0 || includeOptional.size() != 0) {
+        return includeOptional;
```
Comment (wujingya) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Indent one more level

**Review by wujingya — APPROVED** [review]

**Review by MorganAw — APPROVED** [review]

**Pull request merged**